packages/flutter_tools/lib/src/cache.dart

```
@@ -1487,3 +1388,134 @@ const List<List<String>> _dartSdks = <List<String>> [
+    Directory location,
+    void Function(File, Directory) extractor,
+  ) async {
+    final String downloadPath = flattenNameSubdirs(url, _fileSystem);
+    final File tempFile = _createDownloadFile(downloadPath);
+    final Status status = _logger.startProgress(message, timeout: const Duration(minutes: 2));
+    int retries = 0;
+
+    while (retries < 2) {
+      try {
+        _ensureExists(tempFile.parent);
+        await _net.fetchUrl(url, destFile: tempFile, maxAttempts: 2);
+      } finally {
+        status.stop();
+      }
+      _ensureExists(location);
+      try {
+        extractor(tempFile, location);
+      } on ProcessException {
+        if (retries >= 1) {
+          rethrow;
+        }
+        retries += 1;
+        if (tempFile.existsSync()) {
+          tempFile.deleteSync();
+        }
+        continue;
+      }
+      return;
+    }
+  }
+
+  /// Create a temporary file and invoke [onTemporaryFile] with the file as
+  /// argument, then add the temporary file to the [downloadedFiles].
+  File _createDownloadFile(String name) {
+    final File tempFile = _fileSystem.file(_fileSystem.path.join(_tempStorage.path, name));
+    downloadedFiles.add(tempFile);
+    return tempFile;
+  }
+
+  /// Create the given [directory] and parents, as necessary.
+  void _ensureExists(Directory directory) {
+    if (!directory.existsSync()) {
+      directory.createSync(recursive: true);
+    }
+  }
+
+    /// Clear any zip/gzip files downloaded.
+  void removeDownloadedFiles() {
+    for (final File file in downloadedFiles) {
+      if (!file.existsSync()) {
+        continue;
+      }
+      try {
+        file.deleteSync();
+      } on FileSystemException catch (e) {
+        globals.printError('Failed to delete "${file.path}". Please delete manually. $e');
+        continue;
+      }
+      for (Directory d = file.parent; d.absolute.path != _tempStorage.absolute.path; d = d.parent) {
+        if (d.listSync().isEmpty) {
+          d.deleteSync();
```

Ignoring exceptions here too is probably a good idea.